executor/insert_test.go

```diff
@@ -958,3 +959,166 @@ func (s *testSuite3) TestAutoIDIncrementAndOffset(c *C) {
 	c.Assert(err, NotNil)
 	c.Assert(err.Error(), Equals, "[autoid:8060]Invalid auto_increment settings: auto_increment_increment: 65536, auto_increment_offset: 65536, both of them must be in range [1..65535]")
 }
+
+var _ = SerialSuites(&testSuite9{&baseTestSuite{}})
+
+type testSuite9 struct {
+	*baseTestSuite
+}
+
+func (s *testSuite9) TestAutoRandomID(c *C) {
+	allowAutoRandom := config.GetGlobalConfig().Experimental.AllowAutoRandom
+	if !allowAutoRandom {
+		config.GetGlobalConfig().Experimental.AllowAutoRandom = true
+		defer func() {
+			config.GetGlobalConfig().Experimental.AllowAutoRandom = false
+		}()
+	}
+
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec(`use test`)
+	tk.MustExec(`drop table if exists ar`)
+	tk.MustExec(`create table ar (id int key auto_random, name char(10))`)
+
+	tk.MustExec(`insert into ar(id) values (null)`)
+	rs := tk.MustQuery(`select id from ar`)
+	c.Assert(len(rs.Rows()), Equals, 1)
+	firstValue, err := strconv.Atoi(rs.Rows()[0][0].(string))
+	c.Assert(err, IsNil)
+	c.Assert(firstValue, Greater, 0)
+	tk.MustQuery(`select last_insert_id()`).Check(testkit.Rows(fmt.Sprintf("%d", firstValue)))
+	tk.MustExec(`delete from ar`)
+
+	tk.MustExec(`insert into ar(id) values (0)`)
+	rs = tk.MustQuery(`select id from ar`)
+	c.Assert(len(rs.Rows()), Equals, 1)
+	firstValue, err = strconv.Atoi(rs.Rows()[0][0].(string))
+	c.Assert(err, IsNil)
+	c.Assert(firstValue, Greater, 0)
+	tk.MustQuery(`select last_insert_id()`).Check(testkit.Rows(fmt.Sprintf("%d", firstValue)))
+	tk.MustExec(`delete from ar`)
+
+	tk.MustExec(`insert into ar(name) values ('a')`)
+	rs = tk.MustQuery(`select id from ar`)
+	c.Assert(len(rs.Rows()), Equals, 1)
+	firstValue, err = strconv.Atoi(rs.Rows()[0][0].(string))
+	c.Assert(err, IsNil)
+	c.Assert(firstValue, Greater, 0)
+	tk.MustQuery(`select last_insert_id()`).Check(testkit.Rows(fmt.Sprintf("%d", firstValue)))
+
+	tk.MustExec(`drop table ar`)
```

Comment:
how about do rebase first，then check the `last_insert_id`.
eg：insert into ar(id) value(100),(NULL),  the `last_insert_id` should be 101.

Reply:
Situations are too many, we should add many test cases in the future. So I ignore some cases for now.

Reply:
fine.